sys/clif/clif.c

```diff
@@ -0,0 +1,346 @@
+
+    /* if no length given, calculate it */
+    if (!attr->key_len) {
+        attr->key_len = strlen(attr->key);
+    }
+
+    /* count attr name size and separator ';' */
+    size_t req_space = attr->key_len + 1;
+    size_t pos = 0;
+    int quoted = strcmp(attr->key, LF_ATTR_SIZE) ? 1 : 0;
+
+    if (attr->value) {
+        if (!attr->value_len) {
+            attr->value_len = strlen(attr->value);
+        }
+        /* count also '=' */
+        req_space += attr->value_len +  1;
+    }
+
+    if (quoted) {
+        req_space += 2;
+    }
+
+    if (!buf) {
+        return req_space;
+    }
+
+    if (req_space > maxlen) {
+        return CLIF_NO_SPACE;
+    }
+
+    /* add attribute separator ';' */
+    buf[pos++] = LF_ATTR_SEPARATOR_C;
+
+    /* add attribute name */
+    memcpy(&buf[pos], attr->key, attr->key_len);
+    pos += attr->key_len;
+
+    /* add attribute value if defined */
+    if (attr->value) {
+        buf[pos++] = LF_ATTR_VAL_SEPARATOR_C;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+
+        memcpy(&buf[pos], attr->value, attr->value_len);
+        pos += attr->value_len;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+    }
+
+    return pos;
+}
+
+ssize_t clif_get_target(const char *input, size_t input_len, char **output)
+{
+    assert(input);
+    char *target_end;
+
+    *output = memchr(input, LF_PATH_BEGIN_C, input_len);
+    if (!*output) {
+        DEBUG("Path start not found\n");
+        return CLIF_NOT_FOUND;
+    }
+    *output += 1;
+
+    target_end = memchr(*output, LF_PATH_END_C, (input + input_len) - *output);
+    if (!target_end) {
+        DEBUG("Path end not found\n");
+        return CLIF_NOT_FOUND;
+    }
+    ssize_t res = target_end - *output;
+    return res;
+}
+
+ssize_t clif_get_attr(const char *input, size_t input_len, clif_attr_t *attr)
+{
+    assert(input && attr);
+    const char *pos = input;
+    const char *end = input + input_len;
+    unsigned quoted = 0;
+    ssize_t ret_val = CLIF_NOT_FOUND;
+
+    /* initialize attr */
+    attr->value = NULL;
+    attr->key = NULL;
+
+    /* an attribute should start with the separator */
+    if (*pos != LF_ATTR_SEPARATOR_C) {
+        DEBUG("Attribute should start with separator, found %c\n", *pos);
+        goto out;
+    }
+    pos++;
+    attr->key = pos;
+
+    /* iterate over key */
+    while (pos < end) {
+        if (*pos == LF_ATTR_SEPARATOR_C || *pos == LF_LINK_SEPARATOR_C) {
+            /* key ends, no value */
+            attr->key_len = pos - attr->key;
+            goto calculate_out;
+        }
+        if (*pos == LF_ATTR_VAL_SEPARATOR_C) {
+            /* key ends, has value */
+            attr->key_len = pos - attr->key;
+            /* check if the value is quoted and prepare pointer for value scan */
+            pos++;
+            if (*pos == '"') {
+
+                quoted = 1;
+                pos++;
+            }
+            attr->value = pos;
```

Comment:
If you agree to remove const qualifier on the attribute value, then need to cast it away here.